### PR TITLE
Replace multi-select logic with marks

### PR DESCRIPTION
**Pull request by jdfreder**
Mark all the things !

fat marker

all the things

### CODE REVIEW

**Comment by jdfreder** [comment]
@ellisonbg @Carreau

**Comment by ellisonbg** [comment]
Great, will review soon. There are also a few other actions we should enable on marked cells such as run.

**Comment by Carreau** [comment]
No objection. It might just conflict with @ellisonbg action renaming.

File: notebook/static/notebook/js/cell.js [hunk 1]
```diff
@@ -297,7 +293,10 @@ define([
         },
         set: function(value) {
             var isMarked = this.element.hasClass('marked');
-            if (isMarked !== value) {
+            // Use a casting comparison.  Allows for the caller to assign 0 or
+            // 1 instead of a boolean value, which in return means the caller
+            // can do cell.marked ^= true to toggle the mark.
+            if (isMarked != value) {
```
Comment (ellisonbg) [inline notebook/static/notebook/js/cell.js]: LOL, how I love thee JavaScript :)

File: notebook/static/notebook/js/notebook.js [hunk 2]
```diff
@@ -729,29 +714,23 @@ define(function (require) {
...
-        this.get_cell_elements().filter(function (index) {
-            if ($(this).data("cell").in_selection === true) {
-                result.push(index);
-            }
-        });
-        return result;
+    Notebook.prototype.extend_marked = function(direction) {
+        
+        // Convert the direction to an index offset
+        var offset = direction === 'up' ? -1 : 1;
```
Comment (ellisonbg) [inline notebook/static/notebook/js/notebook.js]: After looking at this, I think that `extend_marked` should only accept integers and not `up/down`. Part of this is the difficulty in picking names like `up/down` and `above/below`.

File: notebook/static/notebook/js/notebook.js [hunk 3]
```diff
@@ -771,27 +750,23 @@ define(function (require) {
                 if (this.mode !== 'command') {
                     this.command_mode();
                 }
+                this.get_cell(sindex).unselect();
```
Comment (ellisonbg) [inline notebook/static/notebook/js/notebook.js]: Hmm, this worries me - the logic in select is super subtle. Why is this needed now?

File: notebook/static/notebook/js/notebook.js [hunk 4]
```diff
@@ -771,27 +750,23 @@ define(function (require) {
...
             }
-            var current_selection = this.get_selected_cells();
-            for (var i=0; i<current_selection.length; i++) {
-                current_selection[i].unselect();
+            var cell = this.get_cell(index);
+            cell.select();
+            if (cell.cell_type === 'heading') {
```
Comment (ellisonbg) [inline notebook/static/notebook/js/notebook.js]: Hmm, also don't understand why cell selection needs to fire these events.

Comment (ellisonbg) [inline notebook/static/notebook/js/notebook.js]: IOW is the cell type changing?

File: notebook/static/notebook/js/notebook.js (hunk 3 above)
Comment (jdfreder) [inline notebook/static/notebook/js/notebook.js]: This is the logic that was there before the multiselect logic was added.  I simply reverted the multiselect code in this area.

File: notebook/static/notebook/js/notebook.js (hunk 4 above)
Comment (jdfreder) [inline notebook/static/notebook/js/notebook.js]: Also code that existed before multiselection, just reverted.

**Comment by ellisonbg** [comment]
I have just played with it and overall and really liking it. I do think we should add the marked cell capability for running cells. But I think if there are multiple marked cells the default action should be an inplace run, rather than run and select next. That may change how the keyboard shortcuts and actions for run are desiged. Because of that, let's merge this first without run cell for marked, then merge my command stuff (I will have to rebase) and then submit a final PR to add the marked cell running with the new action names.

But there are a few other things above to fix before this can be merged.

**Comment by ellisonbg** [comment]
Also, I don't mind doing a rebase of my action renaming PR on top of this in its roughly current state.

File: notebook/static/notebook/js/notebook.js (hunk 4 above)
Comment (ellisonbg) [inline notebook/static/notebook/js/notebook.js]: Great! Confusing, but great, and I think we should revert to how you have
it here.

On Fri, Oct 16, 2015 at 1:21 PM, Jonathan Frederic <notifications@github.com
> wrote:

> In notebook/static/notebook/js/notebook.js
> <https://github.com/jupyter/notebook/pull/612#discussion_r42266894>:
>
> >              }
> > -            var current_selection = this.get_selected_cells();
> > -            for (var i=0; i<current_selection.length; i++) {
> > -                current_selection[i].unselect();
> > +            var cell = this.get_cell(index);
> > +            cell.select();
> > +            if (cell.cell_type === 'heading') {
>
> Also code that existed before multiselection, just reverted.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/612/files#r42266894>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by jdfreder** [comment]
@ellisonbg comments addressed

**Comment by ellisonbg** [comment]
OK great, once tests pass I will merge and rebase my action-names PR. Thanks! This is looking really good!

**Comment by ellisonbg** [comment]
Hmmm kicking Travis...

**Comment by jdfreder** [comment]
@ellisonbg Travis passed

**Comment by ellisonbg** [comment]
:)

On Fri, Oct 16, 2015 at 2:51 PM, Matthias Bussonnier <
notifications@github.com> wrote:

> Merged #612 <https://github.com/jupyter/notebook/pull/612>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/612#event-437854750>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com
